src/util/small_map.h

```diff
@@ -0,0 +1,499 @@
+/*******************************************************************\
+
+Module: Small map
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Small map
+
+#ifndef CPROVER_UTIL_SMALL_MAP_H
+#define CPROVER_UTIL_SMALL_MAP_H
+
+#include <bitset>
+#include <utility>
+#include <type_traits>
+#include <memory>
+#include <limits>
+#include <cstdint>
+#include <cstring>
+
+#include "invariant.h"
+
+/// Map from small integers to values
+///
+/// A data structure that maps small integers (in {0, ..., Num-1}) to values.
+/// It is designed to be more memory-efficient than std::map for this use case.
+///
+/// In the following we give some data about the memory consumption of
+/// small_mapt compared to std::map, on a 64-bit Linux system with GNU STL. It
+/// was determined with valgrind/massif. We configured the maps as
+/// `small_mapt<uint64_t, uint32_t, Num=8>` and `std::map<uint8_t, uint64_t>`.
+/// Thus, indices to small_mapt can be in {0, ..., 7} for this configuration.
+///
+/// Below we give the memory consumption (in bytes) of small_mapt and std::map,
+/// both for when containing 8 elements, and when being empty. The value for
+/// "useful" indicates the number of bytes requested, the value for "total" also
+/// includes padding bytes.
+///
+/// small_mapt:
+/// - 8 elements:
+///   * useful:  80 B
+///   * total:  120 B
+/// - empty:
+///   * useful:  16 B
+///   * total:   32 B
+///
+/// std::map:
+/// - 8 elements:
+///   * useful: 432 B
+///   * total:  512 B
+/// - empty:
+///   * useful:  48 B
+///   * total:   64 B
+///
+/// \tparam T mapped type
+/// \tparam Ind unsigned integer type, used to map integer indices to internal
+///   indices that index into the memory block that stores the mapped values
+/// \tparam Num gives range of valid indices, i.e., the valid indices are {0,
+///   ..., Num-1}, must satisfy Num * num_bits(Num-1) + Num < sizeof(Ind) * 8,
+///   with num_bits(n) denoting the minimum number of bits required to represent
+///   integer n
+template<typename T, typename Ind=uint32_t, std::size_t Num=8>
+class small_mapt
+{
+public:
+  static_assert(std::is_unsigned<Ind>::value, "");
+
+  typedef Ind index_fieldt;
+
+  friend void small_map_test();
+
+  small_mapt() : ind(0), p(nullptr) {}
+
+  // Memory
+
+  void copy(T *dest, const T *src, std::size_t n) const
+  {
+    for (std::size_t i = 0; i < n; i++)
+    {
+      new (dest + i) T(*(src + i));
+    }
+  }
+
+  T *allocate(std::size_t n) const
+  {
+    return (T *)malloc(sizeof(T) * n);
```

Comment:
One little thing to be careful of, is that the usage of `malloc` is introducing a very subtle failure case that I think is not being properly handled. If `new` fails, it `throw`s. If `malloc` fails, it returns a null pointer. This can lead to very subtle bugs in `copy` if a machine is stressed in memory. Do you think this is something that might be a good idea to be a `PRECONDITION`, or is it just a waste?

Reply:
Good point. I'm not sure what exactly should happen in this case. I'm gonna check what the STL containers do and then maybe implement it similarly.

Reply:
I think the STL containers use allocators (which in turn use malloc/free) and then use placement new to create the actual objects.

We don't use `new T[n]` here as we want to be able to grow (if possible) and shrink the allocated memory block with `realloc()`. Curiously, C++ allocators don't have an equivalent of `realloc()`. That's why e.g. the capacity of a `std::vector` does not decrease even when erasing elements. C++11 added `shrink_to_fit()`, but that copies the elements to a newly allocated memory segment. Also `new T[n]` would call the default constructor which would be unnecessary when copying objects.

Reply:
So maybe just do `T *ptr = malloc(sizeof(T) * n); if(!ptr) throw std::bad_alloc(); return ptr;` ?

Reply:
Yes, that's the right thing to do.